README
This is a MATE panel applet to display the current uptime. 



ChangeLog
May 15 2019 - Version 0.1.4
===========================
* Fix refresh

Jan 13 2017 - Version 0.1.3
===========================
* Updated translations.
* Added GTK2 autoconf macro.

Jan 7 2016 - Version 0.1.1
===========================
* Updated translations.

Jan 3 2016 - Version 0.1.0
===========================
* Added GTK3 support.

Jan 24 2014 - Version 0.0.5
===========================
* Fixed minute calculation when uptime exceeds 1 day.

Jan 10 2014 - Version 0.0.4
===========================
* Updated menu translations.

Dec 23, 2013 - Version 0.0.3
============================
* Fixed missing XML schema Gettext domain.
* Updated menu labels.

Dec 15, 2013 - Version 0.0.2
============================
* Added Settings menu to chose the display format.
* Added tooltip with full uptime string.

Dec 4, 2013 - Version 0.0.1
============================
* Initial release.



src/main.c
/*
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Library General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, 
 *  USA.
 *
 *  Uptime applet for MATE by Assen Totin <assen.totin@gmail.com>
 *
 */

#include "applet.h"
#include "../config.h"

/* Free the memory of the applet struct
 * Save the current state of the applet
 */
void applet_destroy(MatePanelApplet *applet_widget, UptimeApplet *applet) {
	g_assert(applet);
	g_free(applet);
	return;
}


#ifdef HAVE_GTK2
void applet_back_change (MatePanelApplet *a, MatePanelAppletBackgroundType type, GdkColor *color, GdkPixmap *pixmap, UptimeApplet *applet) {
#elif HAVE_GTK3
void applet_back_change (MatePanelApplet *a, MatePanelAppletBackgroundType type, GdkRGBA *color, cairo_pattern_t *pattern, UptimeApplet *applet) {
#endif

	// Use MATE-provided wrapper to change the background (same for both GTK2 and GTK3)
	mate_panel_applet_set_background_widget (a, GTK_WIDGET(applet->applet));
	//mate_panel_applet_set_background_widget (a, GTK_WIDGET(applet->event_box));
}


/*
 * Get the uptime
 */

void get_uptime(UptimeApplet *applet) {
	FILE *fp = fopen(APPLET_PROC_UPTIME, "r");
	if (!fp)
		return;
	char tmp1[128];
	fgets(&tmp1[0], sizeof(tmp1), fp);
	fclose(fp);
	
	char *tmp2 = strtok(&tmp1[0], " ");
	char *tmp3 = strtok(tmp2, ".");
	applet->uptime = atoi(tmp3);
}


/* 
 * Format uptime
 */
void format_uptime(UptimeApplet *applet, char *s, char *t) {
	int d = applet->uptime / 86400;
	int h = (applet->uptime - d * 86400) / 3600;
	int m = (applet->uptime - (d * 86400) - (h * 3600)) / 60;

	switch(applet->format) {
		case FORMAT_UPPERCASE :
			if (d > 0)
				sprintf(s, "%s%uD%02uH%s", "<span font_desc=\"10.0\">", d, h, "</span>");
			else
				sprintf(s, "%s%02uH%02uM%s", "<span font_desc=\"10.0\">", h, m, "</span>");
			break;
		case FORMAT_LOWERCASE :
			if (d > 0)
				sprintf(s, "%s%ud%02uh%s", "<span font_desc=\"10.0\">", d, h, "</span>");
			else
				sprintf(s, "%s%02uh%02um%s", "<span font_desc=\"10.0\">", h, m, "</span>");
			break;
		case FORMAT_CLOCK :
			sprintf(s, "%s%02u.%02u:%02u%s", "<span font_desc=\"10.0\">", d, h, m, "</span>");
  			break;
		default :
			// Same as FORMAT_UPPERCASE
			if (d > 0)
				sprintf(s, "%s%uD%02uH%s", "<span font_desc=\"10.0\">", d, h, "</span>");
			else
				sprintf(s, "%s%02uH%02uM%s", "<span font_desc=\"10.0\">", h, m, "</span>");
	}

	sprintf(t, "%s %u %s, %u %s, %u %s", _("System uptime:"), d, _("days"), h, _("hours"), m, _("minutes"));
}

/*
 * Update the applet
 */
int applet_check_uptime(UptimeApplet *applet) {
	get_uptime(applet);
	char uptime[128], tooltip[1024]; 

	format_uptime(applet, &uptime[0], &tooltip[0]);
	gtk_label_set_markup(GTK_LABEL(applet->label_bottom), &uptime[0]);
	gtk_widget_set_tooltip_text (GTK_WIDGET (applet->applet), &tooltip[0]);
	
	return TRUE;
}

/* The "main" function
 */
static gboolean uptime_applet_factory(MatePanelApplet *applet_widget, const gchar *iid, gpointer data) {
	char *buf;
	UptimeApplet *applet;
	GdkPixbuf *icon;
	GList *iconlist = NULL;

	if (strcmp (iid, APPLET_ID) != 0) 
		return FALSE;

	// i18n
	setlocale (LC_ALL, "");
	bindtextdomain (PACKAGE_NAME, LOCALEDIR);
	bind_textdomain_codeset(PACKAGE_NAME, "utf-8");
	textdomain (PACKAGE_NAME);

	/* Set an icon for all windows */
/*
	char image_file[1024];
	sprintf(&image_file[0], "%s/%s", APPLET_ICON_PATH, APPLET_ICON);
	icon = gdk_pixbuf_new_from_file(&image_file[0], NULL);
	iconlist = g_list_append(NULL, (gpointer)icon);
	gtk_window_set_default_icon_list(iconlist);
	g_object_unref(icon);
*/	
	applet = g_malloc0(sizeof(UptimeApplet));
	applet->applet = applet_widget;

	applet->gsettings = g_settings_new_with_path(APPLET_GSETTINGS_SCHEMA, APPLET_GSETTINGS_PATH);
	applet->format = g_settings_get_int(applet->gsettings, APPLET_GSETTINGS_KEY_FORMAT);
	
#ifdef HAVE_GTK2
	applet->vbox = gtk_vbox_new (FALSE, 0);
#elif HAVE_GTK3
	applet->vbox = gtk_box_new (GTK_ORIENTATION_VERTICAL, 0);
#endif

	char msg_top[128];
	sprintf(&msg_top[0], "%s%s%s", "<span font_desc=\"8.0\">", _("uptime"), "</span>");
	applet->label_top = gtk_label_new("");
	gtk_label_set_markup(GTK_LABEL(applet->label_top), &msg_top[0]);

	char msg_bottom[128], tooltip[1024];
	get_uptime(applet);
	format_uptime(applet, &msg_bottom[0], &tooltip[0]);
	applet->label_bottom = gtk_label_new("");
	gtk_label_set_markup(GTK_LABEL(applet->label_bottom), &msg_bottom[0]);
	gtk_widget_set_tooltip_text (GTK_WIDGET (applet->applet), &tooltip[0]);

	gtk_box_pack_start(GTK_BOX(applet->vbox), applet->label_top, FALSE, FALSE, 0);
	gtk_box_pack_start(GTK_BOX(applet->vbox), applet->label_bottom, FALSE, FALSE, 0);
	
	gtk_container_add(GTK_CONTAINER(applet_widget), applet->vbox);

	GtkActionGroup *action_group = gtk_action_group_new ("Uptime Applet Actions");
	gtk_action_group_set_translation_domain(action_group, PACKAGE_NAME);
	gtk_action_group_add_actions (action_group, applet_menu_actions, G_N_ELEMENTS (applet_menu_actions), applet);
	mate_panel_applet_setup_menu(applet->applet, ui, action_group);

	gtk_widget_show_all(GTK_WIDGET(applet_widget));

	g_signal_connect(G_OBJECT(applet_widget), "destroy", G_CALLBACK(applet_destroy), (gpointer)applet);
	g_signal_connect(G_OBJECT(applet_widget), "change_background", G_CALLBACK (applet_back_change), (gpointer)applet);

#ifdef HAVE_GTK2
	g_timeout_add(60000, (GtkFunction) applet_check_uptime, (gpointer)applet);
#elif HAVE_GTK3
	g_timeout_add(60000, (GSourceFunc) applet_check_uptime, (gpointer)applet);
#endif

	return TRUE;
}

MATE_PANEL_APPLET_OUT_PROCESS_FACTORY (APPLET_FACTORY, PANEL_TYPE_APPLET, APPLET_NAME, uptime_applet_factory, NULL)





src/menu.c
/*
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Library General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301,
 *  USA.
 *
 *  MATE Uptime applet written by Assen Totin <assen.totin@gmail.com>
 *  
 */

#include "applet.h"
#include "../config.h"


void quitDialogOK( GtkWidget *widget, gpointer data ){
	GtkWidget *quitDialog = data;
	gtk_widget_destroy(quitDialog);
}


void about_cb (GtkAction *action, UptimeApplet *applet) {
	GtkWidget *about = gtk_about_dialog_new();

	gtk_about_dialog_set_program_name (GTK_ABOUT_DIALOG(about), _("Uptime Applet"));

	gtk_about_dialog_set_version (GTK_ABOUT_DIALOG(about), VERSION);

	gtk_about_dialog_set_copyright (GTK_ABOUT_DIALOG(about), "Copyleft 2014-2017. See License for details.");

	gchar *authors[2];
	authors[0] = "Assen Totin <assen.totin@gmail.com>";
	authors[1] = NULL;
	gtk_about_dialog_set_authors (GTK_ABOUT_DIALOG(about),  &authors[0]);

	gtk_about_dialog_set_translator_credits(GTK_ABOUT_DIALOG(about), _("translator-credits"));

	gtk_about_dialog_set_comments (GTK_ABOUT_DIALOG(about), _("An applet that shows the system uptime"));

	gtk_about_dialog_set_website (GTK_ABOUT_DIALOG(about), "http://www.zavedil.com/software-uptime-applet");
	gtk_about_dialog_set_website_label (GTK_ABOUT_DIALOG(about), _("Home page"));

	char image_file[1024];
	snprintf(&image_file[0], 1023, "%s/%s", APPLET_ICON_PATH, "applet_uptime.32.png");
	gtk_about_dialog_set_logo (GTK_ABOUT_DIALOG(about), gtk_image_get_pixbuf(GTK_IMAGE(gtk_image_new_from_file (image_file))));

#ifdef HAVE_GTK2
	gtk_about_dialog_set_license (GTK_ABOUT_DIALOG(about), "GPL v. 2 or later");
#elif HAVE_GTK3
	gtk_about_dialog_set_license_type (GTK_ABOUT_DIALOG(about), GTK_LICENSE_GPL_2_0);
#endif

	gtk_dialog_run (GTK_DIALOG(about));
	gtk_widget_destroy(about);
}


void settings_cb_action(GtkWidget *widget, gpointer data) {
        UptimeApplet *applet = data;
	int format = FORMAT_UPPERCASE;

        gchar *format_text = gtk_combo_box_text_get_active_text(GTK_COMBO_BOX_TEXT(widget));

	if (strstr(format_text, "D"))
		format = FORMAT_UPPERCASE;
	else if (strstr(format_text, "d"))
		format = FORMAT_LOWERCASE;
	else if (strstr(format_text, ":"))
		format = FORMAT_CLOCK;

	applet->format = format;
	g_settings_set_int(applet->gsettings, APPLET_GSETTINGS_KEY_FORMAT, format);

        free(format_text);
}


void settings_cb (GtkAction *action, UptimeApplet *applet) {
	int i;

	// Prepare Feed tab
	GtkWidget *format_label = gtk_label_new(_("Choose output format"));

	GtkWidget *format_combo = gtk_combo_box_text_new();

	gtk_combo_box_text_append_text (GTK_COMBO_BOX_TEXT(format_combo), "04D13H");
	gtk_combo_box_text_append_text (GTK_COMBO_BOX_TEXT(format_combo), "04d13h");
	gtk_combo_box_text_append_text (GTK_COMBO_BOX_TEXT(format_combo), "04.13:53");
	for (i=0; i<NUM_FORMATS; i++) {
		if (applet->format == i) {
			gtk_combo_box_set_active(GTK_COMBO_BOX(format_combo), i);
			break;
		}
	}

	g_signal_connect (G_OBJECT(format_combo), "changed", G_CALLBACK (settings_cb_action), (gpointer) applet);

	GtkWidget *format_vbox_1;
#ifdef HAVE_GTK2
	format_vbox_1 = gtk_vbox_new (FALSE, 0);
#elif HAVE_GTK3
	format_vbox_1 = gtk_box_new (GTK_ORIENTATION_VERTICAL, 0);
#endif

	gtk_box_pack_start(GTK_BOX(format_vbox_1), format_label, FALSE, FALSE, 0);
	gtk_box_pack_start(GTK_BOX(format_vbox_1), format_combo, FALSE, FALSE, 0);

	// Assemble window
	GtkWidget *quitDialog = gtk_dialog_new_with_buttons (_("Uptime Applet Settings"), GTK_WINDOW(applet), GTK_DIALOG_MODAL, NULL);
	GtkWidget *buttonClose = gtk_dialog_add_button (GTK_DIALOG(quitDialog), GTK_STOCK_CLOSE, GTK_RESPONSE_CANCEL);

	gtk_dialog_set_default_response (GTK_DIALOG (quitDialog), GTK_RESPONSE_CANCEL);

#ifdef HAVE_GTK2
	gtk_container_add (GTK_CONTAINER (GTK_DIALOG(quitDialog)->vbox), format_vbox_1);
#elif HAVE_GTK3
	gtk_container_add (GTK_CONTAINER(gtk_dialog_get_content_area(GTK_DIALOG(quitDialog))), format_vbox_1);
#endif

	g_signal_connect (G_OBJECT(buttonClose), "clicked", G_CALLBACK (quitDialogOK), (gpointer) quitDialog);

	gtk_widget_show_all(GTK_WIDGET(quitDialog));
}



src/applet.h
/*
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Library General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301,
 *  USA.
 *
 *  Uptime Applet was writen by Assen Totin <assen.totin@gmail.com>
 *  
 */

#include <mate-panel-applet.h>
#include <libintl.h>
#include <string.h>
#include <stdlib.h>
#include <glib/gi18n.h>

#define APPLET_FACTORY "UptimeAppletFactory"
#define APPLET_ID "UptimeApplet"
#define APPLET_NAME "uptime"
#define APPLET_ICON "applet_uptime.png"
#define APPLET_VERSION "1"

#define APPLET_PROC_UPTIME "/proc/uptime"

// GSettings
#define APPLET_GSETTINGS_SCHEMA "org.mate.panel.applet.UptimeApplet"
#define APPLET_GSETTINGS_PATH "/org/mate/panel/objects/uptime/"
#define APPLET_GSETTINGS_KEY_FORMAT "format"

enum {
	FORMAT_UPPERCASE = 0,
	FORMAT_LOWERCASE,
	FORMAT_CLOCK,
	NUM_FORMATS
};

typedef struct {
	MatePanelApplet *applet;
	//GtkWidget *event_box;
	GtkWidget *vbox, *label_top, *label_bottom;
	int uptime, format;
	GtkDialog *about, *settings;
	GSettings *gsettings;
} UptimeApplet;

void about_cb (GtkAction *, UptimeApplet *);
void settings_cb (GtkAction *, UptimeApplet *);

static const GtkActionEntry applet_menu_actions [] = {
	{ "Properties", GTK_STOCK_PROPERTIES, N_("Settings"), NULL, NULL, G_CALLBACK (settings_cb) },
	{ "About", GTK_STOCK_ABOUT, N_("About"), NULL, NULL, G_CALLBACK (about_cb) }
};

static char *ui  =
"<menuitem name='Item 1' action='Properties' />"
"<menuitem name='Item 2' action='About' />"
;

